File: DP/src/main/python/solution72.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        word1 = " " + word1
        word2 = " " + word2
        m, n = len(word1), len(word2)

        opt = [[float("inf") for _ in range(n)] for _ in range(m)]

        for i in range(m):
            opt[i][0] = i

        for j in range(n):
            opt[0][j] = j

        for i in range(1, m):
            for j in range(1, n):
                if word1[i] == word2[j]:
                    opt[i][j] = min(
                        opt[i - 1][j] + 1, opt[i][j - 1] + 1, opt[i - 1][j - 1]
                    )
                else:
                    opt[i][j] = min(
                        opt[i - 1][j] + 1, opt[i][j - 1] + 1, opt[i - 1][j - 1] + 1
                    )

        # print(opt)
        return opt[m - 1][n - 1]

File: DP/src/main/java/Solution72.java
public class Solution72 {
    public int minDistance(String word1, String word2) {
        word1 = " " + word1;
        word2 = " " + word2;

        int m = word1.length();
        int n = word2.length();

        int[][] opt = new int[m][n];

        for (int i = 0; i < m; i++) {
            opt[i][0] = i;
        }
        for (int j = 0; j < n; j++) {
            opt[0][j] = j;
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                int temp = Math.min(opt[i - 1][j] + 1, opt[i][j - 1] + 1);
                if (word1.charAt(i) == word2.charAt(j)) {
                    opt[i][j] = Math.min(temp, opt[i - 1][j - 1]);
                } else {
                    opt[i][j] = Math.min(temp, opt[i - 1][j - 1] + 1);
                }

            }
        }

        return opt[m - 1][n - 1];
    }
}
